### PR TITLE
Implement ingest source v2

### PR DESCRIPTION
**Pull request by guilload**
### Description
Implement the new ingest source that receives shard assignments from the control plane and streams documents from shards.

Shard assignments from the control plane are a hack at the moment and will be properly implemented in a follow-up PR.

### How was this PR tested?
- Updated ingester unit tests
- Added ingest source v2 unit tests

### CODE REVIEW

**Comment by guilload** [comment]
I have yet to add a bunch of unit tests, but this is ready for review.

File: quickwit/quickwit-indexing/src/source/ingest/mod.rs [hunk 1]
```diff
@@ -0,0 +1,766 @@
...
+impl Source for IngestSource {
+    async fn emit_batches(
+        &mut self,
+        doc_processor_mailbox: &Mailbox<DocProcessor>,
+        ctx: &SourceContext,
+    ) -> Result<Duration, ActorExitStatus> {
+        let now = Instant::now();
+        let mut batch_builder = BatchBuilder::default();
+        let deadline = time::sleep(EMIT_BATCHES_TIMEOUT);
+        tokio::pin!(deadline);
```
Comment (fulmicoton) [inline quickwit/quickwit-indexing/src/source/ingest/mod.rs:193]: tokio::time::time_out does not work?

File: quickwit/quickwit-metastore/src/metastore/file_backed_metastore/file_backed_index/serialize.rs [hunk 2]
```diff
@@ -57,7 +57,7 @@ pub(crate) struct FileBackedIndexV0_6 {
     #[serde(rename = "index")]
     metadata: IndexMetadata,
     splits: Vec<Split>,
-    #[serde(default)]
+    #[serde(default, skip_serializing_if = "HashMap::is_empty")]
```
Comment (fulmicoton) [inline quickwit/quickwit-metastore/src/metastore/file_backed_metastore/file_backed_index/serialize.rs:60]: I'm not sure this is helpful...

File: quickwit/quickwit-proto/src/indexing/mod.rs [hunk 3]
```diff
@@ -153,6 +154,18 @@ impl ToString for IndexingTask {
     }
 }
 
+impl Eq for IndexingTask {}
+
+// TODO: This implementation conflicts with the default derived implementation. It would be better
```
Comment (fulmicoton) [inline quickwit/quickwit-proto/src/indexing/mod.rs:159]: Why does it conflicts? It seems pretty normal?

**Review by fulmicoton — APPROVED** [review]

File: quickwit/quickwit-indexing/src/source/ingest/mod.rs (hunk 1 above)
Comment (guilload) [reply quickwit/quickwit-indexing/src/source/ingest/mod.rs:193]: It does (it's `timeout_at`). All the sources have this defect. I'll fix it when I work on #3829.

File: quickwit/quickwit-proto/src/indexing/mod.rs (hunk 3 above)
Comment (guilload) [reply quickwit/quickwit-proto/src/indexing/mod.rs:159]: The default implementation takes `shard_ids` into account, whereas we only want to consider `index_uid` and `source_id`. I'll improve the comment.

File: quickwit/quickwit-metastore/src/metastore/file_backed_metastore/file_backed_index/serialize.rs (hunk 2 above)
Comment (guilload) [inline quickwit/quickwit-metastore/src/metastore/file_backed_metastore/file_backed_index/serialize.rs:60]: I agree with you and will remove it eventually. I want to hide the shards stuff until it's stable for now.

**Pull request merged**